pkg/psalabelsyncer/sccrolecache.go

```diff
@@ -0,0 +1,440 @@
+package psalabelsyncer
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	rbacv1informers "k8s.io/client-go/informers/rbac/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+
+	securityv1 "github.com/openshift/api/security/v1"
+	securityv1informers "github.com/openshift/client-go/security/informers/externalversions/security/v1"
+	securityv1listers "github.com/openshift/client-go/security/listers/security/v1"
+)
+
+// The index name to be used along with the BySAIndexKeys indexing function
+const BySAIndexName = "ByServiceAccount"
+
+// SAToSCCCache is a construct that caches roles and rolebindings
+// (and their cluster variants) and based on that and on SCCs present in the cluster
+// it allows retrieving a set of SCCs for a given ServiceAccount
+type SAToSCCCache struct {
+	roleLister                rbacv1listers.RoleLister
+	clusterRoleLister         rbacv1listers.ClusterRoleLister
+	roleBindingIndexer        cache.Indexer
+	clusterRoleBindingIndexer cache.Indexer
+
+	sccLister securityv1listers.SecurityContextConstraintsLister
+
+	rolesSynced        cache.InformerSynced
+	roleBindingsSynced cache.InformerSynced
+
+	usefulRolesLock sync.Mutex
+	usefulRoles     map[string][]string
+}
+
+// role and clusterrolebinding object for generic handling, assumes one and
+// at most one of role/clusterrole is always non-nil
+type roleBindingObj struct {
```

Comment:
My preference, given usage, this seems more like

```
type bindingInterface interface{
  RoleRef() rbacv1.RoleRef
  Subjects()
  Namespace()
}
```

With this struct as the impl. Otherwise I fear direct access.